### PR TITLE
Mar 1041

### PR DESCRIPTION
**Pull request by jamespacileo**

### CODE REVIEW

File: templates/barriers/action_plans/add_milestone.html [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+                {% form_field_error form "objective" %}
+
+                <div class="govuk-date-input">
+                    {{ form.objective }}
+                </div>
+            </fieldset>
+        </div>
+        
+        <!-- <div id="{{ form.completion_date.name }}" class="{% form_group_classes form.completion_date.errors %}">
```
Comment (cad106uk) [inline templates/barriers/action_plans/add_milestone.html:36]: Delete commented out code.

File: templates/barriers/action_plans/edit_milestone.html [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+                {% form_field_error form "objective" %}
+
+                <div class="govuk-date-input">
+                    {{ form.objective }}
+                </div>
+            </fieldset>
+        </div>
+        
+        <!-- <div id="{{ form.completion_date.name }}" class="{% form_group_classes form.completion_date.errors %}">
```
Comment (cad106uk) [inline templates/barriers/action_plans/edit_milestone.html:32]: delete not comment out

**Review by cad106uk — CHANGES_REQUESTED** [review]
Why is there the empty file  templates/barriers/action_plans/edit_action_plan.html

File: templates/barriers/action_plans/edit_milestone.html (hunk 2 above)
Comment (jamespacileostint) [reply templates/barriers/action_plans/edit_milestone.html:32]: done

File: templates/barriers/action_plans/add_milestone.html (hunk 1 above)
Comment (jamespacileostint) [reply templates/barriers/action_plans/add_milestone.html:36]: done

File: barriers/views/action_plans.py [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        kwargs["token"] = self.request.session.get("sso_token")
+        kwargs["barrier_id"] = self.kwargs.get("barrier_id")
+        kwargs["milestone_id"] = self.kwargs.get("id")
+        return kwargs
+
+    def get_milestone(self):
+        milestones = self.action_plan.milestones
+        found = list(
```
[the remark this answers is not in the document]
Comment (jamespacileostint) [reply barriers/views/action_plans.py:108]: done

File: barriers/forms/action_plans.py [hunk 4]
```diff
@@ -0,0 +1,430 @@
...
+
+class ActionPlanMilestoneForm(ClearableMixin, APIFormMixin, forms.Form):
+
+    objective = forms.CharField(
+        widget=forms.Textarea(attrs={"class": "govuk-textarea"}),
+        label="Add here what is your milestone objective",
+        error_messages={"required": "Enter your milestone objective"},
+    )
+
+    # completion_date = MonthYearInFutureField(
```
[the remark this answers is not in the document]
Comment (jamespacileostint) [reply barriers/forms/action_plans.py:84]: removed

File: barriers/forms/action_plans.py [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+            milestone_id=self.milestone_id,
+            objective=self.cleaned_data.get("objective"),
+            # completion_date=self.cleaned_data.get("completion_date"),
+        )
+
+    def get_success_url(self):
+        return reverse("barriers:action_plan", kwargs={"barrier_id": self.barrier_id})
+
+
+def get_action_type_category_form(action_type: str):
```
[the remark this answers is not in the document]
Comment (jamespacileo) [reply barriers/forms/action_plans.py:139]: done

File: barriers/forms/action_plans.py [hunk 6]
```diff
@@ -0,0 +1,430 @@
...
+        self.task_id = task_id
+        super().__init__(*args, **kwargs)
+
+    def clean_start_date(self):
+        return self.cleaned_data["start_date"].isoformat()
+
+    def clean_completion_date(self):
+        return self.cleaned_data["completion_date"].isoformat()
+
+    def clean_assigned_to(self):
```
[the remark this answers is not in the document]
Comment (jamespacileo) [reply barriers/forms/action_plans.py:325]: I agree could do with a mixin class I guess, bit burnt out thought eek

File: barriers/views/action_plans.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+from django.views.generic import FormView, TemplateView
+from users.mixins import UserSearchMixin
+from utils.api.client import MarketAccessAPIClient
+
+
+class EditActionPlanCurrentStatusFormView(APIBarrierFormViewMixin, FormView):
+    template_name = "barriers/action_plans/edit_action_plan_current_status.html"
+    form_class = ActionPlanCurrentStatusEditForm
+
+    def get_form_kwargs(self):
```
Comment (cad106uk) [inline barriers/views/action_plans.py:21]: This method should be in a super class

File: barriers/views/action_plans.py [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+        kwargs = super().get_form_kwargs()
+        kwargs["token"] = self.request.session.get("sso_token")
+        kwargs["barrier_id"] = self.kwargs.get("barrier_id")
+        return kwargs
+
+    def get_initial(self):
```
Comment (cad106uk) [inline barriers/views/action_plans.py:27]: This method should be in a super class

File: barriers/views/action_plans.py [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+        if self.request.method == "GET":
+            return self.action_plan.data
+
+    def get_success_url(self):
```
Comment (cad106uk) [inline barriers/views/action_plans.py:31]: This method should be in a super class

File: barriers/views/action_plans.py [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+class SelectActionPlanOwner(BarrierMixin, UserSearchMixin, FormView):
+    template_name = "barriers/action_plans/add_owner.html"
+    error_message = "There was an error adding {full_name} as an owner."
+
+    def select_user_api_call(self, user_id):
+        self.client.action_plans.edit_action_plan(
+            barrier_id=str(self.kwargs.get("barrier_id")), owner=user_id
+        )
+
+    def get_success_url(self):
```
Comment (cad106uk) [inline barriers/views/action_plans.py:46]: This method should be in a super class

File: barriers/views/action_plans.py [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+        return reverse(
+            "barriers:action_plan", kwargs={"barrier_id": self.kwargs.get("barrier_id")}
+        )
+
+
+class AddActionPlanStrategicContext(APIBarrierFormViewMixin, FormView):
+    template_name = "barriers/action_plans/add_strategic_context.html"
+    form_class = ActionPlanStrategicContextForm
+
+    def get_form_kwargs(self):
```
Comment (cad106uk) [inline barriers/views/action_plans.py:56]: This method should be in a super class

File: barriers/views/action_plans.py [hunk 12]
```diff
@@ -0,0 +1,293 @@
...
+        kwargs = super().get_form_kwargs()
+        kwargs["token"] = self.request.session.get("sso_token")
+        kwargs["barrier_id"] = self.kwargs.get("barrier_id")
+        return kwargs
+
+    def get_initial(self):
```
Comment (cad106uk) [inline barriers/views/action_plans.py:62]: This method should be in a super class

File: barriers/views/action_plans.py [hunk 13]
```diff
@@ -0,0 +1,293 @@
...
+        if self.request.method == "GET":
+            return self.action_plan.data
+
+    def get_success_url(self):
```
Comment (cad106uk) [inline barriers/views/action_plans.py:66]: This method should be in a super class

File: barriers/views/action_plans.py [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+        return reverse(
+            "barriers:action_plan", kwargs={"barrier_id": self.kwargs.get("barrier_id")}
+        )
+
+
+class AddActionPlanMilestoneFormView(APIBarrierFormViewMixin, FormView):
+    template_name = "barriers/action_plans/add_milestone.html"
+    form_class = ActionPlanMilestoneForm
+
+    def get_form_kwargs(self):
```
Comment (cad106uk) [inline barriers/views/action_plans.py:76]: This method should be in a super class

File: barriers/views/action_plans.py [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+        kwargs = super().get_form_kwargs()
+        kwargs["token"] = self.request.session.get("sso_token")
+        kwargs["barrier_id"] = self.kwargs.get("barrier_id")
+        return kwargs
+
+    def get_initial(self):
```
Comment (cad106uk) [inline barriers/views/action_plans.py:82]: This method should be in a super class

File: barriers/views/action_plans.py [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+        if self.request.method == "GET":
+            return self.action_plan.data
+
+    def get_success_url(self):
```
Comment (cad106uk) [inline barriers/views/action_plans.py:86]: This method should be in a super class

File: barriers/views/action_plans.py [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+    def get_milestone(self):
+        milestones = self.action_plan.milestones
+        found = [milestone for milestone in milestones if milestone["id"] == str(self.kwargs.get("id"))]
+        return found[0]
+
+    def get_initial(self):
+        if self.request.method == "GET":
+            return self.get_milestone()
+
+    def get_success_url(self):
```
Comment (cad106uk) [inline barriers/views/action_plans.py:112]: This method should be in a super class

File: barriers/views/action_plans.py [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        kwargs["token"] = self.request.session.get("sso_token")
+        kwargs["barrier_id"] = self.kwargs.get("barrier_id")
+        kwargs["action_plan_id"] = self.request.GET.get("action_plan")
+        kwargs["milestone_id"] = self.request.GET.get("milestone")
+        return kwargs
+
+    def get_success_url(self):
```
Comment (cad106uk) [inline barriers/views/action_plans.py:162]: This method should be in a super class

File: barriers/views/action_plans.py [hunk 19]
```diff
@@ -0,0 +1,293 @@
...
+        for milestone in milestones:
+            for task in milestone["tasks"]:
+                if task["id"] == str(self.kwargs.get("id")):
+                    return {**task, "assigned_to": task["assigned_to_email"]}
+
+    def get_initial(self):
+        if self.request.method == "GET":
+            return self.get_task()
+
+    def get_success_url(self):
```
Comment (cad106uk) [inline barriers/views/action_plans.py:192]: This method should be in a super class

File: barriers/views/action_plans.py [hunk 20]
```diff
@@ -0,0 +1,293 @@
...
```
Comment (cad106uk) [inline barriers/views/action_plans.py:222]: This method should be in a super class

File: barriers/views/action_plans.py [hunk 21]
```diff
@@ -0,0 +1,293 @@
...
```
Comment (cad106uk) [inline barriers/views/action_plans.py:252]: This method should be in a super class

**Review by cad106uk — CHANGES_REQUESTED** [review]
There is a lot of duplicate code

**Review by cad106uk — APPROVED** [review]

**Pull request merged**